crates/bevy_reflect/src/set.rs

```diff
@@ -0,0 +1,555 @@
+use std::any::{Any, TypeId};
+use std::fmt::{Debug, Formatter};
+
+use bevy_reflect_derive::impl_type_path;
+use bevy_utils::{Entry, HashMap};
+
+use crate::{
+    self as bevy_reflect, hash_error, ApplyError, Reflect, ReflectKind, ReflectMut, ReflectOwned,
+    ReflectRef, TypeInfo, TypePath, TypePathTable,
+};
+
+/// A trait used to power [set-like] operations via [reflection].
+///
+/// Sets contain zero or more entries of a value, and correspond to types like [`HashSet`]. The
+/// order of these entries is not guaranteed by this trait.
+///
+/// # Hashing
+///
+/// All values are expected to return a valid hash value from [`Reflect::reflect_hash`].
+/// If using the [`#[derive(Reflect)]`](derive@crate::Reflect) macro, this can be done by adding `#[reflect(Hash)]`
+/// to the entire struct or enum.
+/// This is true even for manual implementors who do not use the hashed value,
+/// as it is still relied on by [`DynamicSet`].
+///
+/// # Example
+///
+/// ```
+/// use bevy_reflect::{Reflect, Set};
+/// use bevy_utils::HashSet;
+///
+///
+/// let foo: &mut dyn Set = &mut HashSet::<u32>::new();
+/// foo.insert_boxed(Box::new(123_u32));
+/// assert_eq!(foo.len(), 1);
+///
+/// let field: &dyn Reflect = foo.get(&123_u32).unwrap();
+/// assert_eq!(field.downcast_ref::<u32>(), Some(&123_u32));
+/// ```
+///
+/// [set-like]: https://doc.rust-lang.org/stable/std/collections/struct.HashSet.html
+/// [reflection]: crate
+pub trait Set: Reflect {
+    /// Returns a reference to the value.
+    ///
+    /// If no value is contained, returns `None`.
+    fn get(&self, value: &dyn Reflect) -> Option<&dyn Reflect>;
+
+    /// Returns a mutable reference to the value.
+    ///
+    /// If no value is contained, returns `None`.
+    fn get_mut(&mut self, key: &dyn Reflect) -> Option<&mut dyn Reflect>;
+
+    /// Returns the value at `index` by reference, or `None` if out of bounds.
+    fn get_at(&self, index: usize) -> Option<&dyn Reflect>;
```

Comment:
This doesn't seem like a very useful method for `Set` at large, since sets generally don't guarantee stable iteration order. I feel like this would make more sense just as a plain method on `DynamicSet` specifically; the implementation for "real" sets just uses the `iter` method anyway (so it seems to me like the "right" way to get something like this is to cache the result of calling `iter` and then indexing into it).

Reply:
Seconding this comment.

Reply:
I also feel like this doesn't belong in there. I was mainly just replicating the `Map` trait which also has this. Is there a better explanation for the existence of this method in `Map`?

I'm more than happy to remove this once I find the time to implement the review suggestions!

Reply:
This should also be removed from Map IMO!

Reply:
I wrote something about this [here](https://github.com/bevyengine/bevy/issues/14328#issuecomment-2235735009). I would suggest that we just skip this item of the review change requests and handle it together with the `Map` improvements in the other issue.

Would that be ok with both of you?